### PR TITLE
auto expanding buffer manager

### PR DESCRIPTION
**Pull request by ademar**
ref #379 
ref #363

### CODE REVIEW

**Comment by haf** [comment]
Adding a property to a F# Record is a breaking change. Also, adding a parameter to the `create` function is a breaking change. Either we remove those breaking changes, do them another way that's not breaking or we bump the major version.

**Comment by haf** [comment]
+1 on the functionality :)

**Comment by ademar** [comment]
There are more (possibly breaking) changes coming.

**Comment by haf** [comment]
For the c'tors you can create overloads.

I was thinking we can have a 'v2module' module that has globals for things like `autoGrow` in the configuration & comment out the new configuration field.

When we have half a year worth of these things, we just release v2 and remove the `v2module`

**Comment by ademar** [comment]
Actually I was thinking of scratching the configuration field `autoGrow`, at least for now, and merge these changes under v1 with `autoGrow` enabled.

What do you think?

**Comment by haf** [comment]
I think autoGrow should be on by default, but I still think we should not break public APIs like the c'tors.

File: src/Suave/Sockets/TcpTransport.fs [hunk 1]
```diff
@@ -1,38 +1,43 @@
...
 type TcpTransport(acceptArgs     : SocketAsyncEventArgs,
-                  acceptArgsPool : ConcurrentPool<SocketAsyncEventArgs>,
-                  readArgsPool   : ConcurrentPool<SocketAsyncEventArgs>,
-                  writeArgsPool  : ConcurrentPool<SocketAsyncEventArgs>) =
-  let socket = acceptArgs.AcceptSocket
-  let readArgs = readArgsPool.Pop()
-  let writeArgs = writeArgsPool.Pop()
+                  readArgs     : SocketAsyncEventArgs,
+                  writeArgs     : SocketAsyncEventArgs,
+                  transportPool : ConcurrentPool<TcpTransport>,
```
Comment (haf) [inline src/Suave/Sockets/TcpTransport.fs]: So this is also breaking; can we do something about it?

File: src/Suave/Tcp.fs [hunk 2]
```diff
@@ -32,39 +32,40 @@ let stopTcp (logger : Logger) reason (socket : Socket) =
...
+  acceptArg.add_Completed(fun a b -> userToken.Continuation b)
 
-    readAsyncArgsPool.Push readEventArg
+  new TcpTransport(acceptArg,readEventArg,writeEventArg, transportPool,listenSocket)
 
-    let writeEventArg = new SocketAsyncEventArgs()
-    let userToken = new AsyncUserToken()
-    writeEventArg.UserToken <- userToken
-    writeEventArg.add_Completed(fun a b -> userToken.Continuation b)
+let createPools listenSocket logger maxOps bufferSize autoGrow =
```
Comment (haf) [inline src/Suave/Tcp.fs]: Has changed signature; was this function not public?

File: src/Suave/WebSocket.fs [hunk 3]
```diff
@@ -1,217 +1,217 @@
...
-            do ()
-          | Choice2Of2 err ->
-            Log.log ctx.runtime.logger "Suave.Websocket.handShake" LogLevel.Info (sprintf "websocket disconnected: %A" err)
-          return! Control.CLOSE ctx
-        | _ ->
-          return! RequestErrors.BAD_REQUEST "Missing 'sec-websocket-key' header" ctx
-      | _ ->
-        return! RequestErrors.BAD_REQUEST "Bad Request" ctx
-    }
+namespace Suave
```
Comment (haf) [inline src/Suave/WebSocket.fs]: Can we remove the whitespace changes here?

**Comment by ademar** [comment]
These entities/functions are not intended for public consumption, they are implementation details, we should perhaps mark them internal.  I don't expect anybody would be using any of these anyways.

I have more breaking changes queued towards reducing the number of object allocations on request time. so maybe we could start a branch v2.x with all these changes. or just up the version in master?

What would you do ?

**Comment by haf** [comment]
Up version of master and keep a bugfix branch `releases/v1.x`. Work slightly more with milestones.

**Comment by ademar** [comment]
> Up version of master and keep a bugfix branch releases/v1.x. Work slightly more with milestones.

OK, I'll do that and backport the autogrow logic to v1.1 without modifying the constructors.

**Pull request merged**